config/image-cache.js

```diff
@@ -1,7 +1,94 @@
 #!/usr/bin/env node
-const imgUrl = process.argv[2];
-const encodedUrl = encodeURIComponent(imgUrl);
-const [imgixDomain, imgixToken] = [process.env['IMGIX_DOMAIN'], process.env['IMGIX_TOKEN']];
-const signature = require('crypto').createHash('md5').update(imgixToken + '/' + encodedUrl).digest('hex');
-const resultUrl = new URL(encodedUrl + '?s=' + signature, imgixDomain);
-console.log(resultUrl.toString());
+
+/**
+ * This script is used to upload token metadata images to a Google Cloud Storage bucket. It also
+ * provides the option to resize an image to a max width before uploading so file sizes are more
+ * manageable upon display.
+ *
+ * The following arguments are taken in order from `argv`:
+ * * Remote image URL
+ * * Smart Contract principal
+ * * Token number
+ *
+ * Functionality can be tweaked with the following ENV vars:
+ * * `IMAGE_CACHE_MAX_BYTE_SIZE`: Max payload size accepted when downloading remote images.
+ * * `IMAGE_CACHE_RESIZE_WIDTH`: Width to resize images into while preserving aspect ratio.
+ * * `IMAGE_CACHE_GCS_BUCKET_NAME`: Google Cloud Storage bucket name. Example: 'assets.dev.hiro.so'
+ * * `IMAGE_CACHE_GCS_OBJECT_NAME_PREFIX`: Path for object storage inside the bucket. Example:
+ *   'token-metadata-api/mainnet/'
+ * * `IMAGE_CACHE_GCS_AUTH_TOKEN`: Google Cloud Storage authorization token.
+ * * `IMAGE_CACHE_CDN_BASE_PATH`: Base path for URLs that will be returned to the API for storage.
+ *   Example: 'https://assets.dev.hiro.so/token-metadata-api/mainnet/'
+ */
+
+const sharp = require('sharp');
+const { request, fetch, Agent } = require('undici');
+const { Readable, PassThrough } = require('node:stream');
+
+const IMAGE_URL = process.argv[2];
+const CONTRACT_PRINCIPAL = process.argv[3];
+const TOKEN_NUMBER = process.argv[4];
+
+const IMAGE_RESIZE_WIDTH = parseInt(process.env['IMAGE_CACHE_RESIZE_WIDTH'] ?? '300');
+const GCS_BUCKET_NAME = process.env['IMAGE_CACHE_GCS_BUCKET_NAME'];
+const GCS_OBJECT_NAME_PREFIX = process.env['IMAGE_CACHE_GCS_OBJECT_NAME_PREFIX'];
+const GCS_AUTH_TOKEN = process.env['IMAGE_CACHE_GCS_AUTH_TOKEN'];
+const CDN_BASE_PATH = process.env['IMAGE_CACHE_CDN_BASE_PATH'];
+
+async function upload(stream, name) {
+  try {
+    const response = await request(
+      `https://storage.googleapis.com/upload/storage/v1/b/${GCS_BUCKET_NAME}/o?uploadType=media&name=${GCS_OBJECT_NAME_PREFIX}${name}`,
+      {
+        method: 'POST',
+        body: stream,
+        headers: { 'Content-Type': 'image/png', Authorization: `Bearer ${GCS_AUTH_TOKEN}` },
+      }
+    );
+    if (response.statusCode !== 200) throw new Error(`GCS error: ${response.statusCode}`);
+    return `${CDN_BASE_PATH}${name}`;
+  } catch (error) {
+    throw new Error(`Error uploading ${name}: ${error.message}`);
+  }
+}
+
+fetch(
```

Comment:
There should probably also be a configurable max download size. i.e. prevent the service from downloading giant files (malicious or otherwise)

Reply:
Yep, that's handled by `maxResponseSize` below, in the configured dispatcher

Reply:
Ah I missed that!